jormungandr/src/fragment/pool.rs

```diff
@@ -1,80 +1,54 @@
 use crate::{
     blockcfg::{HeaderContentEvalContext, Ledger, LedgerParameters},
     fragment::{selection::FragmentSelectionAlgorithm, Fragment, Logs},
+    intercom::{NetworkMsg, PropagateMsg},
+    utils::async_msg::MessageBox,
 };
 use chain_core::property::Fragment as _;
-use chain_impl_mockchain::transaction::{AuthenticatedTransaction, Balance};
-use chain_impl_mockchain::value::Value;
+use chain_impl_mockchain::transaction::AuthenticatedTransaction;
 use futures::future::{
     self,
     Either::{A, B},
 };
 use jormungandr_lib::interfaces::{FragmentLog, FragmentOrigin};
+use slog::Logger;
 use std::time::Duration;
 use tokio::{prelude::*, sync::lock::Lock, timer};
 
 #[derive(Clone)]
 pub struct Pool {
     logs: Logs,
     pool: Lock<internal::Pool>,
+    network_msg_box: MessageBox<NetworkMsg>,
 }
 
 impl Pool {
-    pub fn new(ttl: Duration, logs: Logs) -> Self {
+    pub fn new(ttl: Duration, logs: Logs, network_msg_box: MessageBox<NetworkMsg>) -> Self {
         Pool {
             logs,
             pool: Lock::new(internal::Pool::new(ttl)),
+            network_msg_box,
         }
     }
 
     pub fn logs(&self) -> &Logs {
         &self.logs
     }
 
-    /// Returns true if fragment was registered
-    pub fn insert(
-        &mut self,
-        origin: FragmentOrigin,
-        fragment: Fragment,
-    ) -> impl Future<Item = bool, Error = ()> {
-        if !is_fragment_valid(&fragment) {
-            return A(future::ok(false));
-        }
-        let mut pool_lock = self.pool.clone();
-        let mut logs = self.logs.clone();
-        B(self
-            .logs
-            .exists(fragment.id())
-            .and_then(move |exists_in_logs| {
-                if exists_in_logs {
-                    return A(future::ok(false));
-                }
-                B(
-                    future::poll_fn(move || Ok(pool_lock.poll_lock())).and_then(move |mut pool| {
-                        let id = fragment.id();
-                        if pool.insert(fragment) == false {
-                            return A(future::ok(false));
-                        }
-                        let fragment_log = FragmentLog::new(id.into(), origin);
-                        let insert_future = logs.insert(fragment_log).map(|_| true);
-                        B(insert_future)
-                    }),
-                )
-            }))
-    }
-
     /// Returns number of registered fragments
     pub fn insert_all(
```

Comment:
The method name could be changed to indicate that messaging also takes place. `insert_and_propagate_all`?